R/helpers-ppc.R

```diff
@@ -250,6 +276,203 @@ all_counts <- function(x, ...) {
+#'
+#' @param N Length of sample.
+#' @param L Number of chains. Used for MCMC, defaults to 1 for ppc.
+#' @param K Number of equally spaced evaluation points (1:K / K). Defaults to N.
+#' @param prob Desired simultaneous coverage (0,1).
+#' @param M number of simulations to run, if simulation method is used.
+#' @param adj_method String defining the desired adjustment method. By default
+#'  "interpolate" is used. Other available options "optimize" and "simulate".
+#' @return The adjusted coverage parameter yielding the desired simultaneous
+#'  coverage of the ECDF traces.
+#' @noRd
+adjust_gamma <- function(N,
+                         L = 1,
+                         K = N,
+                         prob = 0.99,
+                         M = 1000,
+                         interpolate_adj = FALSE) {
+  if (! all_counts(c(K, N, L))) {
+    abort("Parameters 'N', 'L' and 'K' must be positive integers.")
+  }
+  if (prob >= 1 || prob <= 0) {
+    abort("Value of 'prob' must be in (0,1).")
+  }
+  if (interpolate_adj == TRUE) {
+    gamma <- interpolate_gamma(N, K, prob, L)
+  } else if (L == 1) {
+    gamma <- adjust_gamma_optimize(N, K, prob)
+  } else {
+    gamma <- adjust_gamma_simulate(N,
+                                    L,
+                                    K,
+                                    prob,
+                                    M)
+  }
+  gamma
+}
+
+# Adjust coverage parameter for single sample using the optimization method.
+adjust_gamma_optimize <- function(N, K, prob) {
+  target <- function(gamma, prob, N, K) {
+    z <- 1:(K - 1) / K
+    z1 <- c(0, z)
+    z2 <- c(z, 1)
+
+    # pre-compute quantiles and use symmetry for increased efficiency.
+    x2_lower <- qbinom(gamma / 2, N, z2)
+    x2_upper <- c(N - rev(x2_lower)[2:K], 1)
+
+    # Compute the total probability of trajectories inside the confidence
+    # intervals. Initialize the set and corresponding probabilities known
+    # to be 0 and 1 for the starting value z1 = 0.
+    x1 <- 0
+    p_int <- 1
+    for (i in seq_along(z1)) {
+      tmp <- p_interior(
+        p_int, x1 = x1, x2 = x2_lower[i]: x2_upper[i],
+        z1 = z1[i], z2 = z2[i], N = N
+      )
+      x1 <- tmp$x1
+      p_int <- tmp$p_int
+    }
+    abs(prob - sum(p_int))
+  }
+  optimize(target, c(0, 1 - prob), prob, N = N, K = K)$minimum
+}
+
+# Adjust coverage parameter for multiple chains using simulation method.
+adjust_gamma_simulate <- function(N, L, K, prob, M) {
+  gamma <- numeric(M)
+  z <- (1:(K - 1)) / K
+  n <- N * (L - 1)
+  k <- floor(z * N * L)
+  for (m in seq_len(M)) {
+    u <- u_scale(replicate(L, runif(N)))
+    scaled_ecdfs <- apply(outer(u, z, "<="), c(2, 3), sum)
+    gamma[m] <- 2 * min(
+      apply(
+        scaled_ecdfs, 1, phyper, m = N, n = n, k = k
+      ),
+      apply(
+        scaled_ecdfs - 1, 1, phyper, m = N, n = n, k = k, lower.tail = FALSE
+      )
+    )
+  }
+  alpha_quantile(gamma, 1 - prob)
+}
+
+interpolate_gamma <- function(N, K, p, L) {
```

Comment:
Can you add a bit of internal documentation to these functions? At least an initial line stating what the functions are supposed to do and what they return?

Reply:
I added some documentation to these internal functions. I hope they are now more clear.